### PR TITLE
Add dashboard screen

### PR DESCRIPTION
**Pull request by jopedroliveira**
This PR implements the Station dashboard with listners for saturn activity log,  jobs and earnings.
#119

Identified issues:

- [x] unit tests failing
- [x] jobs counter and activity log not being reactive
- [x] ~~activity log scroll not working properly when window size changes~~

### CODE REVIEW

**Comment by jopedroliveira** [comment]
I think we can start reviewing it.
There are three things missing, two of them can be another PR

- e2e test failing
- activity log keyboard scroll needs to be reviewed in terms of behaviour
- skeleton loading missing

**Comment by juliangruber** [comment]
> I think we can start reviewing it.

please ping when this is ready for review :)

Screenshot 2022-09-26 at 11 10 51

**Comment by juliangruber** [comment]
Got this error in devtools after setting up the wallet:

Screenshot 2022-09-26 at 11 42 20

**Comment by juliangruber** [comment]
Also, I can't scroll the activity log 🤔

**Comment by juliangruber** [comment]
> #75 Must be merged before this one

This has been merged 🥳

File: main/preload.js [hunk 1]
```diff
@@ -52,5 +52,19 @@ contextBridge.exposeInMainWorld('electron', {
     setFilAddress: (/** @type {string | undefined} */ address) => ipcRenderer.invoke('station:setFilAddress', address),
     getOnboardingCompleted: () => ipcRenderer.invoke('station:getOnboardingCompleted'),
     setOnboardingCompleted: () => ipcRenderer.invoke('station:setOnboardingCompleted')
+  },
+  stationEvents: {
+    onActivityLogged: (/** @type {() => Function} */ callback) => {
```
Comment (juliangruber) [inline main/preload.js:57]: should we make the `Function` type more specific?

File: renderer/src/components/ActivityLog.tsx [hunk 2]
```diff
@@ -1,37 +1,72 @@
...
+import { FC } from 'react'
+import { ActivityEventMessage } from '../typings'
+import dayjs from 'dayjs'
+import { ReactComponent as WarningIcon } from '../assets/img/warning.svg'
+import { ReactComponent as JobIcon } from '../assets/img/job.svg'
+
+const ActivityLogItem: FC<ActivityEventMessage> = (activity) => {
+  const time = new Intl.DateTimeFormat(window.navigator.language, {
+    hour: 'numeric', minute: 'numeric', second: 'numeric'
+  }).format(new Date(activity.timestamp))
```
Comment (juliangruber) [inline renderer/src/components/ActivityLog.tsx:10]:
```suggestion

const dateTimeFormat = new Intl.DateTimeFormat(window.navigator.language, {
  hour: 'numeric', minute: 'numeric', second: 'numeric'
})

const ActivityLogItem: FC<ActivityEventMessage> = (activity) => {
  const time = dateTimeFormat.format(new Date(activity.timestamp))
```

Otherwise one could think the formatter needs to be created dynamically

File: renderer/src/lib/station-config.tsx [hunk 3]
```diff
@@ -13,3 +15,40 @@ export async function getFilAddress (): Promise<string | undefined> {
...
+export async function getAllActivities (): Promise<ActivityEventMessage[]> {
+  return await window.electron.getAllActivities()
+}
+
+export async function getTotalEarnings (): Promise<number> {
+  return 0
+}
+
+export async function getTotalJobsCompleted (): Promise<number> {
+  // return 33
```
Comment (juliangruber) [inline renderer/src/lib/station-config.tsx:52]:
```suggestion
```

File: renderer/src/pages/Onboarding.tsx [hunk 4]
```diff
@@ -20,22 +20,24 @@ const OnboardingPage = (): JSX.Element => {
   const navigate = useNavigate()
   const [isLoading, setIsLoading] = useState<boolean>(true)
   const [isOnboardingCompleted, setIsOnboardingCompleted] = useState<boolean|null>()
-
-  useEffect(() => { document.title = 'Filecoin Station' })
+  const [filAddress, setFilAddress] = useState<string|null>()
```
Comment (juliangruber) [inline renderer/src/pages/Onboarding.tsx:23]: interesting, why did this part of onboarding change?

File: renderer/src/pages/WalletConfig.tsx [hunk 5]
```diff
@@ -1,33 +1,18 @@
...
 import { useNavigate } from 'react-router-dom'
-import { getFilAddress, setFilAddress as saveFilAddress } from '../lib/station-config'
+import { setFilAddress as saveFilAddress } from '../lib/station-config'
 
 const WalletConfig = (): JSX.Element => {
   const navigate = useNavigate()
-
-  const [filAddress, setFilAddress] = useState<string | undefined>(undefined)
-
   useEffect(() => { document.title = 'Login' })
```
Comment (juliangruber) [inline renderer/src/pages/WalletConfig.tsx:9]: this should have been removed 🤔  Does this PR have `main` merged in cleanly?

File: test/e2e/app-launch.e2e.test.js [hunk 6]
```diff
@@ -59,24 +59,17 @@ test.describe.serial('Application launch', async () => {
...
-    // https://playwright.dev/docs/api/class-page
-    expect(await mainWindow.title()).toBe('Filecoin Station')
-  })
-
   test('navigate to wallet input', async () => {
     await mainWindow.click('button:has-text("Continue")')
     await mainWindow.click('button:has-text("Continue")')
     await mainWindow.click('button:has-text("Accept")')
-    expect(await mainWindow.title()).toBe('Login')
+    expect(new URL(await mainWindow.url()).pathname).toBe('/wallet')
```
Comment (juliangruber) [inline test/e2e/app-launch.e2e.test.js:66]: this also shouldn't be part of this PR

**Review by juliangruber — CHANGES_REQUESTED** [review]

File: renderer/src/components/ActivityLog.tsx [hunk 7]
```diff
@@ -1,37 +1,72 @@
...
+        return { ...groups, [date]: groups[date] ? groups[date].concat(activity) : [activity] }
+      }, {})
+
+  return (
+    <div className='activity-log'>
+      {Object.keys(activitiesByDate).map((date) => (
+        <div key={date} className="mb-12">
+          <DateSeparator date={date}/>
+          <div>
+            {activitiesByDate[date].map((activity) => <ActivityLogItem key={activity.id} {...activity} />)}
```
Comment (juliangruber) [inline renderer/src/components/ActivityLog.tsx:64]: the spread operator can lead to unepexted results with React component, can we please pass all properties explicitly?

**Review by juliangruber — CHANGES_REQUESTED** [review]

File: renderer/src/pages/Dashboard.tsx [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+  const [totalEarnings, setTotalEarnigs] = useState<number>(0)
+  const [activities, setActivities] = useState<ActivityEventMessage[]>([])
+  const shortAddress = (str: string | undefined) => str
+    ? str.substring(0, 4) + '...' + str.substring(str.length - 4, str.length)
+    : ''
+  const disconnect = async () => {
+    // TODO: move disconnect logic to backend
+    await stopSaturnNode()
+    await setFilAddress('').then(
+      () => { setAddress(undefined); navigate('/', { replace: true }) }
```
Comment (juliangruber) [inline renderer/src/pages/Dashboard.tsx:27]:
```suggestion
      () => { setAddress(undefined); navigate('/wallet', { replace: true }) }
```

This shouldn't go back to onboarding

Comment (juliangruber) [reply renderer/src/pages/Dashboard.tsx:27]: Also, could you please refactor not to use `.then()`?

**Review by juliangruber — CHANGES_REQUESTED** [review]

**Review by juliangruber — CHANGES_REQUESTED** [review]

File: renderer/src/pages/Dashboard.tsx [hunk 9]
```diff
@@ -35,29 +35,30 @@ const Dashboard = (): JSX.Element => {
...
+    setTotalJobs(value)
   }, [])
 
-  useEffect(() => { document.title = 'Dashboard' })
-
   useEffect(() => {
     reload()
-    startSaturnNode()
+    // startSaturnNode()
+    // return stopSaturnNode()
```
Comment (jopedroliveira) [inline renderer/src/pages/Dashboard.tsx:53]: @juliangruber I'm not sure if this should be done here in the dashboard, to ensure station starts and stops running on mount and on dismount. Somehow, it's related with`window.electron.saturnNode.start()` on `WalletConfig.tsx`. What do you think?

File: main/preload.js (hunk 1 above)
Comment (jopedroliveira) [reply main/preload.js]: [e31bfe8](https://github.com/filecoin-project/filecoin-station/pull/144/commits/e31bfe884190e67c2ad3073b71126d5f64332567)

File: renderer/src/pages/WalletConfig.tsx (hunk 5 above)
Comment (jopedroliveira) [reply renderer/src/pages/WalletConfig.tsx]: [bbd4874](https://github.com/filecoin-project/filecoin-station/pull/144/commits/bbd4874fb7d2c3334f0ba0cca3eaaba70f97b899)

File: renderer/src/pages/Onboarding.tsx (hunk 4 above)
Comment (jopedroliveira) [reply renderer/src/pages/Onboarding.tsx]: [bbd4874](https://github.com/filecoin-project/filecoin-station/pull/144/commits/bbd4874fb7d2c3334f0ba0cca3eaaba70f97b899)

File: renderer/src/components/ActivityLog.tsx (hunk 7 above)
Comment (jopedroliveira) [reply renderer/src/components/ActivityLog.tsx]: [6800270](https://github.com/filecoin-project/filecoin-station/pull/144/commits/68002702b9462d2925937c413ca7e406441a231b)

[the remark this answers is not in the document]
Comment (jopedroliveira) [reply renderer/src/components/ActivityLog.tsx]: [6800270](https://github.com/filecoin-project/filecoin-station/pull/144/commits/68002702b9462d2925937c413ca7e406441a231b)

File: renderer/src/pages/Dashboard.tsx (hunk 9 above)
Comment (juliangruber) [reply renderer/src/pages/Dashboard.tsx:53]: If we change `WalletConfig.tsx` to only start Saturn when the address changes, and start Saturn initially when the `App` gets mounted (as long as there's an address configured), I think we won't have any extra start events. I'm going to PR it to this branch.

File: renderer/src/pages/Dashboard.tsx [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+  const handleActivity = useCallback((value: ActivityEventMessage[]) => {
+    setActivities(value)
+  }, [])
+
+  const handleEarnings = useCallback((value: number) => {
+    setTotalEarnigs(value)
+  }, [])
+
+  const handleJobs = useCallback((value: number) => {
```
Comment (juliangruber) [inline renderer/src/pages/Dashboard.tsx:46]: these shouldn't be necessary, you can just pass `setTotalJobs` directly, right?

Comment (jopedroliveira) [reply renderer/src/pages/Dashboard.tsx:46]: Currently yes, but I'd prefer to explicitly define a callback to handle such recalls. One in the future may want to process it other way or trigger anything else when this event is triggered.
Anyway, If you prefer the implicit way, I can change

File: renderer/src/pages/Dashboard.tsx (hunk 9 above)
Comment (juliangruber) [reply renderer/src/pages/Dashboard.tsx]: I committed it right to this branch in https://github.com/filecoin-project/filecoin-station/pull/144/commits/1720b0fab52cc37dea44e93b8f6eb64ebdc750a0, as I can't file PRs against your fork unfortunately

File: renderer/src/pages/Dashboard.tsx (hunk 10 above)
Comment (juliangruber) [reply renderer/src/pages/Dashboard.tsx:46]: > One in the future may want to process it other way or trigger anything else when this event is triggered.

If that is the case then it's easy to update. We can keep the `useCallbacks` while this PR is still WIP, so you can iterate quickly should you need to change these, but let's please clean these up when the PR is done

Comment (jopedroliveira) [reply renderer/src/pages/Dashboard.tsx]: [afd5e98](https://github.com/filecoin-project/filecoin-station/pull/144/commits/afd5e98b0223258d90a6658c64dd3aea2af46f8a)

File: main/preload.js [hunk 11]
```diff
@@ -52,5 +52,26 @@ contextBridge.exposeInMainWorld('electron', {
     setFilAddress: (/** @type {string | undefined} */ address) => ipcRenderer.invoke('station:setFilAddress', address),
     getOnboardingCompleted: () => ipcRenderer.invoke('station:getOnboardingCompleted'),
     setOnboardingCompleted: () => ipcRenderer.invoke('station:setOnboardingCompleted')
+  },
+  stationEvents: {
+    onActivityLogged: (/** @type {(value: import('./typings').Activity) => void} */ callback) => {
```
[the remark this answers is not in the document]
Comment (juliangruber) [reply main/preload.js:57]: see for example `before-pack.js`

File: renderer/src/pages/Dashboard.tsx [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+
+  const reload = async (): Promise<void> => {
+    setAddress(await getFilAddress())
+    setActivities(await getAllActivities())
+    setTotalEarnigs(await getTotalEarnings())
+    setTotalJobs(await getTotalJobsCompleted())
+  }
+
+  useEffect(() => {
+    reload()
```
Comment (juliangruber) [inline renderer/src/pages/Dashboard.tsx:39]: this method is only called by one place, can we inline it into this effect?

**Review by juliangruber — CHANGES_REQUESTED** [review]

**Comment by juliangruber** [comment]
could you please update the PR description to reflect that #75 has landed?

File: renderer/src/lib/station-config.tsx [hunk 13]
```diff
@@ -13,3 +15,39 @@ export async function getFilAddress (): Promise<string | undefined> {
...
+  return await window.electron.saturnNode.getLog()
+}
+
+export async function stopSaturnNode (): Promise<void> {
+  return await window.electron.saturnNode.stop()
+}
+
+export async function startSaturnNode (): Promise<void> {
+  return await window.electron.saturnNode.start()
+}
```
[the remark this answers is not in the document]
Comment (jopedroliveira) [reply renderer/src/lib/station-config.tsx:41]: Sure. It would make more sense to add it on #171

File: renderer/src/pages/Dashboard.tsx [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+  const [address, setAddress] = useState<string | undefined>()
+  const [totalJobs, setTotalJobs] = useState<number>(0)
+  const [totalEarnings, setTotalEarnigs] = useState<number>(0)
+  const [activities, setActivities] = useState<ActivityEventMessage[]>([])
+  const shortAddress = (str: string | undefined) => str
+    ? str.substring(0, 4) + '...' + str.substring(str.length - 4, str.length)
+    : ''
+  const disconnect = async () => {
+    // TODO: move disconnect logic to backend
```
[the remark this answers is not in the document]
Comment (jopedroliveira) [reply renderer/src/pages/Dashboard.tsx:24]: The point here is if we want to trigger `stopSaturnNode` here. The alternative would be to expose an API responsible to stop Saturn (and other modules) and disconnect the wallet all at once, instead of making it explicitly in a frontend page. Thoughts??

File: renderer/src/pages/Dashboard.tsx [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+import { useNavigate } from 'react-router-dom'
+
+const Dashboard = (): JSX.Element => {
+  const navigate = useNavigate()
+
+  const [address, setAddress] = useState<string | undefined>()
+  const [totalJobs, setTotalJobs] = useState<number>(0)
+  const [totalEarnings, setTotalEarnigs] = useState<number>(0)
+  const [activities, setActivities] = useState<ActivityEventMessage[]>([])
+  const shortAddress = (str: string | undefined) => str
```
[the remark this answers is not in the document]
Comment (jopedroliveira) [reply renderer/src/pages/Dashboard.tsx]: [afe985f](https://github.com/filecoin-project/filecoin-station/pull/144/commits/afe985f2c16844cd75fc02a4e40fa4734b8b61b0)

File: renderer/src/lib/station-config.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (jopedroliveira) [reply renderer/src/lib/station-config.tsx:41]: [870068b](https://github.com/filecoin-project/filecoin-station/pull/171/commits/870068b765216394aa2e3213e1bd582c7442607f)

File: renderer/src/pages/Dashboard.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (juliangruber) [reply renderer/src/pages/Dashboard.tsx]: As long as there's no problem doing this in the frontend, I don't see why we should change it. We do own the frontend 100% so there's no risk (like with web services) where the backend must be in control of all the business logic

**Review by juliangruber — COMMENTED** [review]
this kicks some serious ass

**Review by juliangruber — APPROVED** [review]
